From a0d97de67282c2aa36a4397575e0639a69c61311 Mon Sep 17 00:00:00 2001
From: Sven Schneider <sven.schneider@h-brs.de>
Date: Tue, 28 May 2024 12:06:39 +0200
Subject: [PATCH 2/2] fix multiple type-scoped contexts

The previous version only expanded terms that have been defined in the first suitable type.
This version iterates over all types and merges their respective contexts in a single dictionary.
Then that dictionary is the basis for the construction of a subcontext.
---
 rdflib/plugins/shared/jsonld/context.py | 19 +++++++++++++------
 1 file changed, 13 insertions(+), 6 deletions(-)

diff --git a/rdflib/plugins/shared/jsonld/context.py b/rdflib/plugins/shared/jsonld/context.py
index 1eec9441..d5ea89a7 100644
--- a/rdflib/plugins/shared/jsonld/context.py
+++ b/rdflib/plugins/shared/jsonld/context.py
@@ -158,18 +158,25 @@ class Context:
             if not isinstance(rtype, list):
                 rtype = [rtype] if rtype else []
 
-            typeterm = None
-            for rt in rtype:
+            typeterm_context = {}
+            # iterate in reverse direction so that the first types have higher priority
+            for rt in reversed(rtype):
                 try:
                     typeterm = self.terms.get(rt)
                 except TypeError:
                     # extra lenience, triggers if type is set to a literal
                     pass
-                if typeterm is not None:
-                    break
 
-            if typeterm and typeterm.context:
-                subcontext = self.subcontext(typeterm.context, propagate=False)
+                if typeterm and typeterm.context:
+                    # contexts can either be dictionaries or lists
+                    if isinstance(typeterm.context, dict):
+                        typeterm_context |= typeterm.context
+                    else:
+                        for context in typeterm.context:
+                            typeterm_context |= context
+
+            if typeterm_context:
+                subcontext = self.subcontext(typeterm_context, propagate=False)
                 if subcontext:
                     return subcontext
 
-- 
2.48.0

